protocols/autonatv2/src/server/handler/dial_request.rs

```diff
@@ -0,0 +1,277 @@
+use std::{
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::future::FusedFuture;
+use futures::{
+    channel::{mpsc, oneshot},
+    AsyncRead, AsyncWrite, AsyncWriteExt, SinkExt, StreamExt,
+};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, ListenUpgradeError},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use rand_core::RngCore;
+
+use crate::request_response::Coder;
+use crate::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    request_response::{
+        DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+    },
+    Nonce, REQUEST_UPGRADE,
+};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum DialBack {
+    #[allow(unused)]
+    Dial,
+    DialBack,
+    Ok,
+}
```

Comment:
I don't understand what this does. Can we improve the naming here?

Reply:
I changed the name a bit and added doc comments. I hope it's clearer now.